src/distPlus/symbolic/SymbolicDist.re

```diff
@@ -296,103 +313,407 @@ module GenericSimple = {
+      cutShape;
+    };
+
+    switch (et) {
+      | `Simple(d) => {
+          let (sc, sd) = renderDistributionToXYShape(d, n);
+
+          let scc = sc |> Distributions.Continuous.shapeMap(cut);
+          let sdc = sd |> cut;
+
+          let newIntegral = 1.; // TODO
+
+          `RenderedShape(scc, sdc, newIntegral);
+      }
+      | `RenderedShape(sc, sd, i) => {
+          let scc = sc |> Distributions.Continuous.shapeMap(cut);
+          let sdc = sd |> cut;
+
+          let newIntegral = 1.; // TODO
+
+          `RenderedShape(scc, sdc, newIntegral);
+      }
+    }
+  };
+
+  let evaluateVerticalScaling = (et1: nodeResult, et2: nodeResult, n: int) => {
+    let scale = (i: float, s: DistTypes.xyShape): DistTypes.xyShape => {xs: s.xs, ys: s.ys |> E.A.fmap(y => y *. i)};
+
+    switch ((et1, et2)) {
+      | (`Simple(`Float(v)), `Simple(d))
+      | (`Simple(d), `Simple(`Float(v))) => {
+          let (sc, sd) = renderDistributionToXYShape(d, n);
+
+          let scc = sc |> Distributions.Continuous.shapeMap(scale(v));
+          let sdc = sd |> scale(v);
+
+          let newIntegral = v; // TODO
+
+          `RenderedShape(scc, sdc, newIntegral);
+      }
+      | (`Simple(`Float(v)), `RenderedShape(sc, sd, i))
+      | (`RenderedShape(sc, sd, i), `Simple(`Float(v))) => {
+          let scc = sc |> Distributions.Continuous.shapeMap(scale(v));
+          let sdc = sd |> scale(v);
+
+          let newIntegral = v; // TODO
+
+          `RenderedShape(scc, sdc, newIntegral);
+        }
+      | _ => `RenderedShape(Distributions.Continuous.empty, Distributions.Discrete.empty, 0.) // TODO: give warning
+    }
+  }
+
+  let renderNode = (et: nodeResult, n: int) => {
+    switch (et) {
+    | `Simple(d) => {
+          let (sc, sd) = renderDistributionToXYShape(d, n);
+          `RenderedShape(sc, sd, 1.0);
+      }
+    | s => s
+    }
+  }
+
+  let rec evaluateNode = (treeNode: distTree, n: int): nodeResult => {
+    // returns either a new symbolic distribution
+    switch (treeNode) {
+    | `Simple(d) => evaluateDistribution(d)
+    | `Combination(t1, t2, op) => evaluateCombinationDistribution(evaluateNode(t1, n), evaluateNode(t2, n), op, n)
+    | `PointwiseSum(t1, t2) => evaluatePointwiseSum(evaluateNode(t1, n), evaluateNode(t2, n), n)
+    | `PointwiseProduct(t1, t2) => evaluatePointwiseProduct(evaluateNode(t1, n), evaluateNode(t2, n), n)
+    | `VerticalScaling(t1, t2) => evaluateVerticalScaling(evaluateNode(t1, n), evaluateNode(t2, n), n)
+    | `Normalize(t) => evaluateNormalize(evaluateNode(t, n), n)
+    | `LeftTruncate(t, x) => evaluateTruncate(evaluateNode(t, n), x, (>=), n)
+    | `RightTruncate(t, x) => evaluateTruncate(evaluateNode(t, n), x, (<=), n)
+    | `Render(t) => renderNode(evaluateNode(t, n), n)
+    }
+  };
+
+  let toShape = (treeNode: distTree, n: int) => {
+    let treeShape = evaluateNode(`Render(`Normalize(treeNode)), n);
+
+    switch (treeShape) {
+    | `Simple(_) => E.O.toExn("No shape found!", None)
```

Comment:
Note that this is an antipattern, though we do use it sometimes to iterate quickly. Generally functional languages should return optionals or Results, rather than raising errors.

Reply:
Good point. `renderNode` should return ``result(`RenderedShape)``, and that will fix that